File: PersianCalendar/src/main/java/net/androgames/level/view/LevelView.java
package net.androgames.level.view;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.core.content.res.ResourcesCompat;

import com.byagowi.persiancalendar.R;

import net.androgames.level.orientation.Orientation;

import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.Locale;

import static com.byagowi.persiancalendar.utils.UtilsKt.a11yAnnounceAndClick;

/*
 *  This file is part of Level (an Android Bubble Level).
 *  <https://github.com/avianey/Level>
 *
 *  Copyright (C) 2014 Antoine Vianey
 *
 *  Level is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Level is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Level. If not, see <http://www.gnu.org/licenses/>
 */
public class LevelView extends View {
    private static final double LEVEL_ASPECT_RATIO = 0.150;
    private static final double BUBBLE_WIDTH = 0.150;
    private static final double BUBBLE_ASPECT_RATIO = 1.000;
    private static final double BUBBLE_CROPPING = 0.500;
    private static final double MARKER_GAP = BUBBLE_WIDTH + 0.020;
    /**
     * Angle max
     */
    private static final double MAX_SINUS = Math.sin(Math.PI / 4);
    /**
     * Fonts and colors
     */
    private static final String FONT_LCD = "fonts/lcd.ttf";
    private boolean isAlreadyLeveled = true; // deliberately
    /**
     * Dimensions
     */
    private int canvasWidth;
    private int canvasHeight;
    private int minLevelX;
    private int maxLevelX;
    private int levelWidth;
    private int levelHeight;
    private int levelMinusBubbleWidth;
    private int levelMinusBubbleHeight;
    private int middleX;
    private int middleY;
    private int halfBubbleWidth;
    private int halfBubbleHeight;
    private int halfMarkerGap;
    private int minLevelY;
    private int maxLevelY;
    private int minBubble;
    private int maxBubble;
    private int markerThickness;
    private int levelBorderWidth;
    private int levelBorderHeight;
    private int lcdWidth;
    private int lcdHeight;
    private int displayPadding;
    private int displayGap;
    private int sensorGap;
    private int levelMaxDimension;
    /**
     * Rect
     */
    private final Rect displayRect = new Rect();
    /**
     * Angles
     */
    private float angle1;
    private float angle2;
    /**
     * Orientation
     */
    private Orientation orientation;
    private long lastTime;
    private long lastTimeShowAngle;
    private double angleToShow1;
    private double angleToShow2;
    private double angleX;
    private double angleY;
    private double speedX;
    private double speedY;
    private double x, y;
    /**
     * Drawables
     */
    private Drawable level1D;
    private Drawable bubble1D;
    private Drawable marker1D;
    private Drawable level2D;
    private Drawable bubble2D;
    private Drawable marker2D;
    private Drawable display;
    /**
     * Ajustement de la vitesse
     */
    private double viscosityValue = 1;
    /**
     * Format des angles
     */
    private final DecimalFormat displayFormat = new DecimalFormat("00.0");
    private static final String displayBackgroundText = "88.8";
    private final Paint lcdForegroundPaint = new Paint();
    private final Paint lcdBackgroundPaint = new Paint();
    private final Paint infoPaint = new Paint();
    private boolean firstTime = true;

    public LevelView(Context context) {
        super(context);
        init(context);
    }

    public LevelView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    private void init(Context context) {
        Resources resources = context.getResources();
        // drawable
        level1D = ResourcesCompat.getDrawable(resources, R.drawable.level_1d, null);
        level2D = ResourcesCompat.getDrawable(resources, R.drawable.level_2d, null);
        bubble1D = ResourcesCompat.getDrawable(resources, R.drawable.bubble_1d, null);
        bubble2D = ResourcesCompat.getDrawable(resources, R.drawable.bubble_2d, null);
        marker1D = ResourcesCompat.getDrawable(resources, R.drawable.marker_1d, null);
        marker2D = ResourcesCompat.getDrawable(resources, R.drawable.marker_2d, null);
        display = ResourcesCompat.getDrawable(resources, R.drawable.display, null);

        // config
        displayFormat.setDecimalFormatSymbols(new DecimalFormatSymbols(Locale.ENGLISH));

        // typeface
        Typeface lcd = Typeface.createFromAsset(context.getAssets(), FONT_LCD);

        // paint
        infoPaint.setColor(resources.getColor(R.color.black));
        infoPaint.setAntiAlias(true);
        infoPaint.setTextSize(resources.getDimensionPixelSize(R.dimen.info_text));
        infoPaint.setTypeface(Typeface.create(Typeface.SANS_SERIF, Typeface.NORMAL));
        infoPaint.setTextAlign(Paint.Align.CENTER);

        lcdForegroundPaint.setColor(resources.getColor(R.color.lcd_front));
        lcdForegroundPaint.setAntiAlias(true);
        lcdForegroundPaint.setTextSize(resources.getDimensionPixelSize(R.dimen.lcd_text));
        lcdForegroundPaint.setTypeface(lcd);
        lcdForegroundPaint.setTextAlign(Paint.Align.CENTER);

        lcdBackgroundPaint.setColor(resources.getColor(R.color.lcd_back));
        lcdBackgroundPaint.setAntiAlias(true);
        lcdBackgroundPaint.setTextSize(resources.getDimensionPixelSize(R.dimen.lcd_text));
        lcdBackgroundPaint.setTypeface(lcd);
        lcdBackgroundPaint.setTextAlign(Paint.Align.CENTER);

        // dimens
        lcdBackgroundPaint.getTextBounds(displayBackgroundText, 0, displayBackgroundText.length(), displayRect);
        lcdHeight = displayRect.height();
        lcdWidth = displayRect.width();
        levelBorderWidth = resources.getDimensionPixelSize(R.dimen.level_border_width);
        levelBorderHeight = resources.getDimensionPixelSize(R.dimen.level_border_height);
        markerThickness = resources.getDimensionPixelSize(R.dimen.marker_thickness);
        displayGap = resources.getDimensionPixelSize(R.dimen.display_gap);
        sensorGap = resources.getDimensionPixelSize(R.dimen.sensor_gap);
        displayPadding = resources.getDimensionPixelSize(R.dimen.display_padding);

        invalidate();
    }

    public void setOrientation(Orientation newOrientation, float newPitch, float newRoll, float newBalance) {
        if (orientation == null || !orientation.equals(newOrientation)) {
            orientation = newOrientation;

            int infoY;
            switch (newOrientation) {
                case LEFT:      // left
                case RIGHT:     // right
                    infoY = (canvasHeight - canvasWidth) / 2 + canvasWidth;
                    break;
                case TOP:       // top
                case BOTTOM:    // bottom
                default:        // landing
                    infoY = canvasHeight;
                    break;
            }

            int sensorY = infoY - sensorGap;

            middleX = canvasWidth / 2;
            middleY = canvasHeight / 2;

            // level
            switch (newOrientation) {
                case LANDING:   // landing
                    levelWidth = levelMaxDimension;
                    levelHeight = levelMaxDimension;
                    break;
                case TOP:       // top
                case BOTTOM:    // bottom
                case LEFT:      // left
                case RIGHT:     // right
                    levelWidth = canvasWidth - 2 * displayGap;
                    levelHeight = (int) (levelWidth * LEVEL_ASPECT_RATIO);
                    break;
            }

            viscosityValue = levelWidth;

            minLevelX = middleX - levelWidth / 2;
            maxLevelX = middleX + levelWidth / 2;
            minLevelY = middleY - levelHeight / 2;
            maxLevelY = middleY + levelHeight / 2;

            // bubble
            halfBubbleWidth = (int) (levelWidth * BUBBLE_WIDTH / 2);
            halfBubbleHeight = (int) (halfBubbleWidth * BUBBLE_ASPECT_RATIO);
            int bubbleWidth = 2 * halfBubbleWidth;
            int bubbleHeight = 2 * halfBubbleHeight;
            maxBubble = (int) (maxLevelY - bubbleHeight * BUBBLE_CROPPING);
            minBubble = maxBubble - bubbleHeight;

            // display
            displayRect.set(
                    middleX - lcdWidth / 2 - displayPadding,
                    sensorY - displayGap - 2 * displayPadding - lcdHeight,
                    middleX + lcdWidth / 2 + displayPadding,
                    sensorY - displayGap);

            // marker
            halfMarkerGap = (int) (levelWidth * MARKER_GAP / 2);

            // autres
            levelMinusBubbleWidth = levelWidth - bubbleWidth - 2 * levelBorderWidth;
            levelMinusBubbleHeight = levelHeight - bubbleHeight - 2 * levelBorderWidth;

            // positionnement
            level1D.setBounds(minLevelX, minLevelY, maxLevelX, maxLevelY);
            level2D.setBounds(minLevelX, minLevelY, maxLevelX, maxLevelY);
            marker2D.setBounds(
                    middleX - halfMarkerGap - markerThickness,
                    middleY - halfMarkerGap - markerThickness,
                    middleX + halfMarkerGap + markerThickness,
                    middleY + halfMarkerGap + markerThickness);

            x = ((double) (maxLevelX + minLevelX)) / 2;
            y = ((double) (maxLevelY + minLevelY)) / 2;
        }
        switch (orientation) {
            case TOP:
            case BOTTOM:
                angle1 = Math.abs(newBalance);
                angleX = Math.sin(Math.toRadians(newBalance)) / MAX_SINUS;
                break;
            case LANDING:
                angle2 = Math.abs(newRoll);
                angleX = Math.sin(Math.toRadians(newRoll)) / MAX_SINUS;
            case RIGHT:
            case LEFT:
                angle1 = Math.abs(newPitch);
                angleY = Math.sin(Math.toRadians(newPitch)) / MAX_SINUS;
                if (angle1 > 90) {
                    angle1 = 180 - angle1;
                }
                break;
        }
        // correction des angles affiches
        if (angle1 > 99.9f) {
            angle1 = 99.9f;
        }
        if (angle2 > 99.9f) {
            angle2 = 99.9f;
        }
        // correction des angles aberrants
        // pour ne pas que la bulle sorte de l'ecran
        if (angleX > 1) {
            angleX = 1;
        } else if (angleX < -1) {
            angleX = -1;
        }
        if (angleY > 1) {
            angleY = 1;
        } else if (angleY < -1) {
            angleY = -1;
        }
        // correction des angles a plat
        // la bulle ne doit pas sortir du niveau
        if (orientation.equals(Orientation.LANDING) && angleX != 0 && angleY != 0) {
            double n = Math.sqrt(angleX * angleX + angleY * angleY);
            double teta = Math.acos(Math.abs(angleX) / n);
            double l = 1 / Math.max(Math.abs(Math.cos(teta)), Math.abs(Math.sin(teta)));
            angleX = angleX / l;
            angleY = angleY / l;
        }

        if (orientation.isLevel(newPitch, newRoll, newBalance, .8f)) {
            if (!isAlreadyLeveled) {
                a11yAnnounceAndClick(this, R.string.level);
                isAlreadyLeveled = true;
            }
        } else {
            isAlreadyLeveled = false;
        }

        postInvalidate();
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);

        canvasWidth = w;
        canvasHeight = h;

        levelMaxDimension = Math.min(
                Math.min(h, w) - 2 * displayGap,
                Math.max(h, w) - 2 * (sensorGap + 3 * displayGap + lcdHeight));
    }

    @Override
    public void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        canvas.save();

        if (firstTime) {
            setOrientation(Orientation.LANDING, 0, 0, 0);
            firstTime = false;
        }

        // update physics
        long currentTime = System.currentTimeMillis();
        double timeDiff = (currentTime - lastTime) / 1000.0;
        lastTime = currentTime;

        if (currentTime - lastTimeShowAngle > 500) {
            angleToShow1 = angle1;
            angleToShow2 = angle2;
            lastTimeShowAngle = currentTime;
        }

        double posX = orientation.getReverse() * (2 * x - minLevelX - maxLevelX) / levelMinusBubbleWidth;
        switch (orientation) {
            case TOP:
            case BOTTOM:
                speedX = orientation.getReverse() * (angleX - posX) * viscosityValue;
                break;
            case LEFT:
            case RIGHT:
                speedX = orientation.getReverse() * (angleY - posX) * viscosityValue;
                break;
            case LANDING:
                double posY = (2 * y - minLevelY - maxLevelY) / levelMinusBubbleHeight;
                speedX = (angleX - posX) * viscosityValue;
                speedY = (angleY - posY) * viscosityValue;
                y += speedY * timeDiff;
                break;
        }
        x += speedX * timeDiff;
        // en cas de latence elevee
        // si la bubble a trop deviee
        // elle est replacee correctement
        if (orientation == Orientation.LANDING) {
            if (Math.sqrt((middleX - x) * (middleX - x)
                    + (middleY - y) * (middleY - y)) > levelMaxDimension / 2 - halfBubbleWidth) {
                x = (angleX * levelMinusBubbleWidth + minLevelX + maxLevelX) / 2;
                y = (angleY * levelMinusBubbleHeight + minLevelY + maxLevelY) / 2;
            }
        } else {
            if (x < minLevelX + halfBubbleWidth || x > maxLevelX - halfBubbleWidth) {
                x = (angleX * levelMinusBubbleWidth + minLevelX + maxLevelX) / 2;
            }
        }

        if (orientation == Orientation.LANDING) {
            // Angle
            display.setBounds(
                    displayRect.left - (displayRect.width() + displayGap) / 2,
                    displayRect.top,
                    displayRect.right - (displayRect.width() + displayGap) / 2,
                    displayRect.bottom);
            display.draw(canvas);
            display.setBounds(
                    displayRect.left + (displayRect.width() + displayGap) / 2,
                    displayRect.top,
                    displayRect.right + (displayRect.width() + displayGap) / 2,
                    displayRect.bottom);
            display.draw(canvas);
            canvas.drawText(
                    displayBackgroundText,
                    middleX - (displayRect.width() + displayGap) / 2,
                    displayRect.centerY() + lcdHeight / 2,
                    lcdBackgroundPaint);
            canvas.drawText(
                    displayFormat.format(angleToShow2),
                    middleX - (displayRect.width() + displayGap) / 2,
                    displayRect.centerY() + lcdHeight / 2,
                    lcdForegroundPaint);
            canvas.drawText(
                    displayBackgroundText,
                    middleX + (displayRect.width() + displayGap) / 2,
                    displayRect.centerY() + lcdHeight / 2,
                    lcdBackgroundPaint);
            canvas.drawText(
                    displayFormat.format(angleToShow1),
                    middleX + (displayRect.width() + displayGap) / 2,
                    displayRect.centerY() + lcdHeight / 2,
                    lcdForegroundPaint);
            //
            bubble2D.setBounds(
                    (int) (x - halfBubbleWidth),
                    (int) (y - halfBubbleHeight),
                    (int) (x + halfBubbleWidth),
                    (int) (y + halfBubbleHeight));
            level2D.draw(canvas);
            bubble2D.draw(canvas);
            marker2D.draw(canvas);
            canvas.drawLine(minLevelX, middleY,
                    middleX - halfMarkerGap, middleY, infoPaint);
            canvas.drawLine(middleX + halfMarkerGap, middleY,
                    maxLevelX, middleY, infoPaint);
            canvas.drawLine(middleX, minLevelY,
                    middleX, middleY - halfMarkerGap, infoPaint);
            canvas.drawLine(middleX, middleY + halfMarkerGap,
                    middleX, maxLevelY, infoPaint);
        } else {
            canvas.rotate(orientation.getRotation(), middleX, middleY);
            // Angle
            display.setBounds(displayRect);
            display.draw(canvas);
            canvas.drawText(
                    displayBackgroundText, middleX,
                    displayRect.centerY() + lcdHeight / 2, lcdBackgroundPaint);
            canvas.drawText(
                    displayFormat.format(angleToShow1), middleX,
                    displayRect.centerY() + lcdHeight / 2, lcdForegroundPaint);
            // level
            level1D.draw(canvas);
            // bubble
            canvas.clipRect(
                    minLevelX + levelBorderWidth, minLevelY + levelBorderHeight,
                    maxLevelX - levelBorderWidth, maxLevelY - levelBorderHeight);
            bubble1D.setBounds(
                    (int) (x - halfBubbleWidth), minBubble,
                    (int) (x + halfBubbleWidth), maxBubble);
            bubble1D.draw(canvas);
            // marker
            marker1D.setBounds(
                    middleX - halfMarkerGap - markerThickness, minLevelY,
                    middleX - halfMarkerGap, maxLevelY);
            marker1D.draw(canvas);
            marker1D.setBounds(
                    middleX + halfMarkerGap, minLevelY,
                    middleX + halfMarkerGap + markerThickness, maxLevelY);
            marker1D.draw(canvas);
        }

        canvas.restore();
    }
}

File: PersianCalendar/src/main/kotlin/com/byagowi/persiancalendar/ui/level/LevelView.kt
/*
 * This file was originally a part of Level (an Android Bubble Level).
 * <https://github.com/avianey/Level>
 *
 * Copyright (C) 2014 Antoine Vianey
 *
 * Level is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Level is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Level. If not, see <http://www.gnu.org/licenses/>
 */
package net.androgames.level

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import androidx.core.graphics.withRotation
import com.byagowi.persiancalendar.R
import com.byagowi.persiancalendar.ui.utils.dp
import com.byagowi.persiancalendar.ui.utils.getCompatDrawable
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.acos
import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin

class LevelView(context: Context, attrs: AttributeSet? = null) : View(context, attrs) {

    private val angleDisplay = AngleDisplay(context)
    private val infoPaint = Paint(Paint.ANTI_ALIAS_FLAG).also {
        it.color = Color.BLACK
    }

    /**
     * Dimensions
     */
    private var canvasWidth = 0
    private var canvasHeight = 0
    private var minLevelX = 0
    private var maxLevelX = 0
    private var levelWidth = 0
    private var levelHeight = 0
    private var levelMinusBubbleWidth = 0
    private var levelMinusBubbleHeight = 0
    private var middleX = 0
    private var middleY = 0
    private var halfBubbleWidth = 0
    private var halfBubbleHeight = 0
    private var halfMarkerGap = 0
    private var minLevelY = 0
    private var maxLevelY = 0
    private var minBubble = 0
    private var maxBubble = 0
    private val markerThickness = resources.getDimensionPixelSize(R.dimen.marker_thickness)
    private val levelBorderWidth = resources.getDimensionPixelSize(R.dimen.level_border_width)
    private val levelBorderHeight = resources.getDimensionPixelSize(R.dimen.level_border_height)
    private val sensorGap = 5.dp.toInt()
    private var levelMaxDimension = 0

    /**
     * Angles
     */
    private var angle1 = 0f
    private var angle2 = 0f

    /**
     * Orientation
     */
    private var orientation: Orientation? = null
    private var lastTime: Long = 0
    private var lastTimeShowAngle: Long = 0
    private var angleToShow1 = 0f
    private var angleToShow2 = 0f
    private var angleX = 0.0
    private var angleY = 0.0
    private var speedX = 0.0
    private var speedY = 0.0
    private var x = 0.0
    private var y = 0.0

    /**
     * Drawables
     */
    private val level1D = context.getCompatDrawable(R.drawable.level_1d)
    private val bubble1D = context.getCompatDrawable(R.drawable.bubble_1d)
    private val marker1D = context.getCompatDrawable(R.drawable.marker_1d)
    private val level2D = context.getCompatDrawable(R.drawable.level_2d)
    private val bubble2D = context.getCompatDrawable(R.drawable.bubble_2d)
    private val marker2D = context.getCompatDrawable(R.drawable.marker_2d)

    /**
     * Ajustement de la vitesse
     */
    private var viscosityValue = 1.0
    private var firstTime = true

    fun setOrientation(
        newOrientation: Orientation, newPitch: Float, newRoll: Float, newBalance: Float
    ) {
        if (orientation == null || orientation != newOrientation) {
            orientation = newOrientation
            middleX = canvasWidth / 2
            middleY = canvasHeight / 2 - angleDisplay.displayGap
            when (newOrientation) {
                Orientation.LANDING -> {
                    levelWidth = levelMaxDimension
                    levelHeight = levelMaxDimension
                }
                Orientation.TOP, Orientation.BOTTOM, Orientation.LEFT, Orientation.RIGHT -> {
                    levelWidth = canvasWidth - 2 * angleDisplay.displayGap
                    levelHeight = (levelWidth * LEVEL_ASPECT_RATIO).toInt()
                }
            }
            viscosityValue = levelWidth.toDouble()
            minLevelX = middleX - levelWidth / 2
            maxLevelX = middleX + levelWidth / 2
            minLevelY = middleY - levelHeight / 2
            maxLevelY = middleY + levelHeight / 2

            // bubble
            halfBubbleWidth = (levelWidth * BUBBLE_WIDTH / 2).toInt()
            halfBubbleHeight = (halfBubbleWidth * BUBBLE_ASPECT_RATIO).toInt()
            val bubbleWidth = 2 * halfBubbleWidth
            val bubbleHeight = 2 * halfBubbleHeight
            maxBubble = (maxLevelY - bubbleHeight * BUBBLE_CROPPING).toInt()
            minBubble = maxBubble - bubbleHeight

            // display
            val displayY = when (newOrientation) {
                Orientation.LEFT, Orientation.RIGHT ->
                    (canvasHeight - canvasWidth) / 2 + canvasWidth - angleDisplay.displayGap
                else -> canvasHeight
            }
            angleDisplay.updatePlacement(middleX, displayY)

            // marker
            halfMarkerGap = (levelWidth * MARKER_GAP / 2).toInt()

            // autres
            levelMinusBubbleWidth = levelWidth - bubbleWidth - 2 * levelBorderWidth
            levelMinusBubbleHeight = levelHeight - bubbleHeight - 2 * levelBorderWidth

            // positionnement
            level1D.setBounds(minLevelX, minLevelY, maxLevelX, maxLevelY)
            level2D.setBounds(minLevelX, minLevelY, maxLevelX, maxLevelY)
            marker2D.setBounds(
                middleX - halfMarkerGap - markerThickness,
                middleY - halfMarkerGap - markerThickness,
                middleX + halfMarkerGap + markerThickness,
                middleY + halfMarkerGap + markerThickness
            )
            x = (maxLevelX + minLevelX).toDouble() / 2
            y = (maxLevelY + minLevelY).toDouble() / 2
        }
        when (orientation) {
            Orientation.TOP, Orientation.BOTTOM -> {
                angle1 = abs(newBalance)
                angleX = sin(Math.toRadians(newBalance.toDouble())) / MAX_SINUS
            }
            Orientation.LANDING -> {
                angle2 = abs(newRoll)
                angleX = sin(Math.toRadians(newRoll.toDouble())) / MAX_SINUS
                angle1 = abs(newPitch)
                angleY = sin(Math.toRadians(newPitch.toDouble())) / MAX_SINUS
                if (angle1 > 90) {
                    angle1 = 180 - angle1
                }
            }
            Orientation.RIGHT, Orientation.LEFT -> {
                angle1 = abs(newPitch)
                angleY = sin(Math.toRadians(newPitch.toDouble())) / MAX_SINUS
                if (angle1 > 90) {
                    angle1 = 180 - angle1
                }
            }
            else -> Unit
        }
        // correction des angles affiches
        angle1 = angle1.coerceAtMost(99.9f)
        angle2 = angle2.coerceAtMost(99.9f)
        // correction des angles aberrants
        // pour ne pas que la bulle sorte de l'ecran
        angleX = angleX.coerceIn(-1.0, 1.0)
        angleY = angleY.coerceIn(-1.0, 1.0)
        // correction des angles a plat
        // la bulle ne doit pas sortir du niveau
        if (orientation == Orientation.LANDING && angleX != 0.0 && angleY != 0.0) {
            val n = hypot(angleX, angleY)
            val teta = acos(abs(angleX) / n)
            val l = 1 / max(abs(cos(teta)), abs(sin(teta)))
            angleX /= l
            angleY /= l
        }
        onIsLevel(newOrientation.isLevel(newPitch, newRoll, newBalance, .8f))
        invalidate()
    }

    var onIsLevel = fun(_: Boolean) {}

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        canvasWidth = w
        canvasHeight = h
        levelMaxDimension = min(
            min(h, w) - 2 * angleDisplay.displayGap,
            max(h, w) - 2 * (sensorGap + 3 * angleDisplay.displayGap + angleDisplay.lcdHeight)
        )
        angleDisplay.updatePlacement(canvasWidth / 2, canvasHeight)
    }

    public override fun onDraw(canvas: Canvas) {
        if (firstTime) {
            setOrientation(Orientation.LANDING, 0f, 0f, 0f)
            firstTime = false
        }
        val orientation = orientation ?: return

        // update physics
        val currentTime = System.currentTimeMillis()
        val timeDiff = (currentTime - lastTime) / 1000.0
        lastTime = currentTime
        if (currentTime - lastTimeShowAngle > 500) {
            angleToShow1 = angle1
            angleToShow2 = angle2
            lastTimeShowAngle = currentTime
        }
        val posX = orientation.reverse * (2 * x - minLevelX - maxLevelX) / levelMinusBubbleWidth
        when (orientation) {
            Orientation.TOP, Orientation.BOTTOM ->
                speedX = orientation.reverse * (angleX - posX) * viscosityValue
            Orientation.LEFT, Orientation.RIGHT ->
                speedX = orientation.reverse * (angleY - posX) * viscosityValue
            Orientation.LANDING -> {
                val posY = (2 * y - minLevelY - maxLevelY) / levelMinusBubbleHeight
                speedX = (angleX - posX) * viscosityValue
                speedY = (angleY - posY) * viscosityValue
                y += speedY * timeDiff
            }
        }
        x += speedX * timeDiff
        // en cas de latence elevee
        // si la bubble a trop deviee
        // elle est replacee correctement
        if (orientation == Orientation.LANDING) {
            if (hypot(middleX - x, middleY - y) > levelMaxDimension / 2 - halfBubbleWidth) {
                x = (angleX * levelMinusBubbleWidth + minLevelX + maxLevelX) / 2
                y = (angleY * levelMinusBubbleHeight + minLevelY + maxLevelY) / 2
            }
        } else {
            if (x < minLevelX + halfBubbleWidth || x > maxLevelX - halfBubbleWidth) {
                x = (angleX * levelMinusBubbleWidth + minLevelX + maxLevelX) / 2
            }
        }
        if (orientation == Orientation.LANDING) {
            bubble2D.setBounds(
                (x - halfBubbleWidth).toInt(),
                (y - halfBubbleHeight).toInt(),
                (x + halfBubbleWidth).toInt(),
                (y + halfBubbleHeight).toInt()
            )
            level2D.draw(canvas)
            bubble2D.draw(canvas)
            marker2D.draw(canvas)
            canvas.drawLine(
                minLevelX.toFloat(), middleY.toFloat(),
                (middleX - halfMarkerGap).toFloat(), middleY.toFloat(), infoPaint
            )
            canvas.drawLine(
                (middleX + halfMarkerGap).toFloat(), middleY.toFloat(),
                maxLevelX.toFloat(), middleY.toFloat(), infoPaint
            )
            canvas.drawLine(
                middleX.toFloat(), minLevelY.toFloat(),
                middleX.toFloat(), (middleY - halfMarkerGap).toFloat(), infoPaint
            )
            canvas.drawLine(
                middleX.toFloat(), (middleY + halfMarkerGap).toFloat(),
                middleX.toFloat(), maxLevelY.toFloat(), infoPaint
            )

            angleDisplay.draw(canvas, angleToShow1, offsetXFactor = -1)
            angleDisplay.draw(canvas, angleToShow2, offsetXFactor = 1)
        } else canvas.withRotation(
            orientation.rotation.toFloat(), middleX.toFloat(), middleY.toFloat()
        ) {
            angleDisplay.draw(canvas, angleToShow1)
            // level
            level1D.draw(canvas)
            // bubble
            canvas.clipRect(
                minLevelX + levelBorderWidth, minLevelY + levelBorderHeight,
                maxLevelX - levelBorderWidth, maxLevelY - levelBorderHeight
            )
            bubble1D.setBounds(
                (x - halfBubbleWidth).toInt(), minBubble,
                (x + halfBubbleWidth).toInt(), maxBubble
            )
            bubble1D.draw(canvas)
            // marker
            marker1D.setBounds(
                middleX - halfMarkerGap - markerThickness, minLevelY,
                middleX - halfMarkerGap, maxLevelY
            )
            marker1D.draw(canvas)
            marker1D.setBounds(
                middleX + halfMarkerGap, minLevelY,
                middleX + halfMarkerGap + markerThickness, maxLevelY
            )
            marker1D.draw(canvas)
        }
    }

    companion object {
        private const val LEVEL_ASPECT_RATIO = 0.150
        private const val BUBBLE_WIDTH = 0.150
        private const val BUBBLE_ASPECT_RATIO = 1.000
        private const val BUBBLE_CROPPING = 0.500
        private const val MARKER_GAP = BUBBLE_WIDTH + 0.020
        private val MAX_SINUS = sin(PI / 4) // Angle max
    }
}
